docs: add RELEASING.md handbook (#11026)
tauri-apps/tauri: .github/RELEASING.md

# Tauri Releasing Handbook

This handbook contains information about our release pipeline and how to deal with common issues.
This document is mainly intended for team members responsible for maintaining the project.

- [Covector](#covector)
- [Version Pull Request](#version-pull-request)
- [Releasing and Publishing](#releasing-and-publishing)
- [Publishing failed, what to do?](#publishing-failed-what-to-do)

> # Covector

Reviewer: ```suggestion
## Covector
```

We use [`covector`](https://github.com/jbolda/covector) to manage our version bumps and release pipeline.
It can be configured in [`.changes/config.json`](../.changes/config.json) which includes how each package should be published step by step.

Some packages can't be published directly using `covector` as it requires to be built on a matrix of platforms

> such as `tauri-cli` prebuilt binraies which is published using [publish-cli-rs.yml](./workflows/publish-cli-rs.yml)

Reviewer 2: ```suggestion
such as `tauri-cli` prebuilt binaries which is published using [publish-cli-rs.yml](./workflows/publish-cli-rs.yml)
```

and `@tauri-apps/cli` native Node.js modules which is published using using [publish-cli-js.yml](./workflows/publish-cli-js.yml)
both of which are triggered after `covector` has created a github release for both of them, see `Trigger @tauri-apps/cli publishing workflow`
and `Trigger tauri-cli publishing workflow` steps in [covector-version-or-publish.yml](./workflows/covector-version-or-publish.yml)

> # Version Pull Request

Reviewer: ```suggestion
## Version Pull Request
```

On each pull request merged, [covector-version-or-publish.yml](./workflows/covector-version-or-publish.yml) workflow will run, and:

When there're change files inside `.changes` folder and they're not all included in `pre.json` (usually this is only when we are in `-alpha` to `-rc` phase), it will open/update an `Apply Version Updates From Current Changes` PR (https://github.com/tauri-apps/tauri/pull/11029 for example) that bumps all packages based on current existing change files and generate `CHANGELOG.md` entries. see `Create Pull Request With Versions Bumped` step in [covector-version-or-publish.yml](./workflows/covector-version-or-publish.yml).

Otherwise, covector will start to publish packages configured in [`.changes/config.json`](../.changes/config.json).

> # Releasing and Publishing

Reviewer: ```suggestion
## Releasing and Publishing
```

Releasing can be as easy as merging the version pull request but here is a checklist to follow:

- [ ] Double check that every package is bumped correctly and there are no accidental major or minor being released unless that is indeed the intention.
- [ ] Make sure that there are no pending or unfinished [covector-version-or-publish.yml](./workflows/covector-version-or-publish.yml) workflow runs.
- [ ] Sign the Version PR before merging as we require signed commits
  - [ ] `git fetch --all`
  - [ ] `git checkout release/version-updates`
  - [ ] `git commit --amend -S`
  - [ ] `git push --force`
- [ ] Approve and merge the version pull request

> # Publishing failed, what to do?

Reviewer: ```suggestion
## Publishing failed, what to do?
```

# Tauri Releasing Handbook

This handbook contains information about our release pipeline and how to deal with common issues.
This document is mainly intended for team members responsible for maintaining the project.

- [Covector](#covector)
- [Version Pull Request](#version-pull-request)
- [Releasing and Publishing](#releasing-and-publishing)
- [Publishing failed, what to do?](#publishing-failed-what-to-do)

# Covector

> We use [`covector`](https://github.com/jbolda/covector) to manage our version pumbs and relase pipeline.

Reviewer: ```suggestion
We use [`covector`](https://github.com/jbolda/covector) to manage our version bumps and release pipeline.
```

It can be configured in [`.changes/config.json`](../.changes/config.json) which includes how each package should be published step by step.

Some packages can't be published directly using `covector` as it requires to be built on a matrix of platforms

> such as `tauri-cli` prebuilt binraies which is published using using [publish-cli-rs.yml](./workflows/publish-cli-rs.yml)

Reviewer: ```suggestion
such as `tauri-cli` prebuilt binraies which is published using [publish-cli-rs.yml](./workflows/publish-cli-rs.yml)
```

and `@tauri-apps/cli` native Node.js modules which is published using using [publish-cli-js.yml](./workflows/publish-cli-js.yml)
both of which are triggered after `covector` has created a github release for both of them, see `Trigger @tauri-apps/cli publishing workflow`
and `Trigger tauri-cli publishing workflow` steps in [covector-version-or-publish.yml](./workflows/covector-version-or-publish.yml)

# Version Pull Request

> On each pull request merged, [covector-version-or-publish.yml](./workflows/covector-version-or-publish.yml) workflow will run to determine whether to publish or not

Reviewer: This confused me a lot to be honest, I need to work my way backwards to understand that this is basically saying when you remove all the change files, it will trigger the publish process

Author: Sorry about that, I am not very good at writing, how would you rephrase it?

Reviewer: I'll try, but I'm not good at it either 😂

> # Version Pull Request
> 
> On each pull request merged, [covector-version-or-publish.yml](./workflows/covector-version-or-publish.yml) workflow will run to determine whether to publish or not
> and it determines that based on two conditions:
> 
> - `.changes` doesn't have any change files in it
> - `.changes` has change files but they all included in `pre.json` (usually this is only when we are in `-alpha` to `-rc` phase).
> 
> If one of these conditions are met, covector will start to publish packages configured in [`.changes/config.json`](../.changes/config.json).
> 
> Otherwise, it will bump all packages based on current existing change files and generate CHANGELOG.md entries. These changes will then be included in a version PR, see `Create Pull Request With Versions Bumped` step in [covector-version-or-publish.yml](./workflows/covector-version-or-publish.yml).

Reviewer: ```suggestion
# Version Pull Request

On each pull request merged, [covector-version-or-publish.yml](./workflows/covector-version-or-publish.yml) workflow will run, and:

When there're change files inside `.changes` folder and they're not all included in `pre.json` (usually this is only when we are in `-alpha` to `-rc` phase), it will open/update an `Apply Version Updates From Current Changes` PR (https://github.com/tauri-apps/tauri/pull/11029 for example) that bumps all packages based on current existing change files and generate `CHANGELOG.md` entries. see `Create Pull Request With Versions Bumped` step in [covector-version-or-publish.yml](./workflows/covector-version-or-publish.yml).

Otherwise, covector will start to publish packages configured in [`.changes/config.json`](../.changes/config.json).
```

This is probably the best I can do 😂

I'm gonna get some sleep, I can get back to this tomorrow

Reviewer 3: sleep is for the weak 😠

Author: > This is probably the best I can do 

this is good enough

> Otherwise, it will pump all packages based on current existing change files and generate CHANGELOG.md entries. These changes will then be included in a version PR, see `Create Pull Request With Versions Bumped` step in [covector-version-or-publish.yml](./workflows/covector-version-or-publish.yml).

Reviewer: ```suggestion
Otherwise, it will bump all packages based on current existing change files and generate CHANGELOG.md entries. These changes will then be included in a version PR, see `Create Pull Request With Versions Bumped` step in [covector-version-or-publish.yml](./workflows/covector-version-or-publish.yml).
```

# Releasing and Publishing

Releasing can be as easy as merging the version pull request but here is a checklist to follow:

> - [ ] Double check that every package is pumped correctly and there is no accidental major or minor being released unless that is indeed the intention.

Reviewer: ```suggestion
- [ ] Double check that every package is bumped correctly and there are no accidental major or minor being released unless that is indeed the intention.
```

> - [ ] Make sure that there is no pending or unfinished [covector-version-or-publish.yml](./workflows/covector-version-or-publish.yml) workflow runs.

Reviewer: ```suggestion
- [ ] Make sure that there are no pending or unfinished [covector-version-or-publish.yml](./workflows/covector-version-or-publish.yml) workflow runs.
```

> - [ ] Sign the Version PR before merging as we require signed commits

Reviewer: Can the bot sign the commit so that we don't need to do this ourselves?

https://github.com/peter-evans/create-pull-request/releases/tag/v7.0.0 seems be able to do it

Author: We are not using [peter-evans/create-pull-request](https://github.com/peter-evans/create-pull-request) but it is not impossible

ref: https://github.com/jbolda/covector/issues/272

I wouldn't hold this PR on it though

Reviewer: Saying this is because we're using https://github.com/tauri-apps/create-pull-request which is a fork of it, and since I don't see any changes in the fork, I don't know why we did it to be honest

> I wouldn't hold this PR on it though

Sure, I didn't know we need to do this before, just pointing it out

Author: > Saying this is because we're using [tauri-apps/create-pull-request](https://github.com/tauri-apps/create-pull-request) which is a fork of it, and since I don't see any changes in the fork, I don't know why we did it to be honest

I believe it was for security reasons, cc @tweidinger @nothingismagick

Reviewer 4: Happened before I joined and it hasn't been maintained or updated by us, so I would rather delete our fork and use upstream with a version tag/commit hash.

  - [ ] `git fetch --all`
  - [ ] `git checkout release/version-updates`
  - [ ] `git commit --amend -S`
  - [ ] `git push --force`
- [ ] Approve and merge the version pull request

# Publishing failed, what to do?

It is possible and due to many factors that one or many packages release can fail to release, there is no reason to panic, we can fix this.

> Did all of the packages failed to release?

Reviewer: ```suggestion
Did all of the packages fail to release?
```
